### PR TITLE
DAOS-11232 tests: Add a vos command line test

### PR DESCRIPTION
**Pull request by jolivier23**
I used this basic test to help me debug the issues with aggregation and found it generically useful for trying various scenarios so I'm committing the code. Also, it adds some regression testing for the ticket in question

Required-githooks: true

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

**Review by NiuYawei — DISMISSED** [review]

**Comment by jolivier23** [comment]
ping reviewers

File: src/vos/tests/vos_cmd.c [hunk 1]
```diff
@@ -0,0 +1,1064 @@
...
+	int rc;
+
+	rc = ABT_init(0, NULL);
+	if (rc != ABT_SUCCESS) {
+		fprintf(stderr, "ABT init failed: %d\n", rc);
+		return -1;
+	}
+
+	rc = ABT_xstream_self(&abt_xstream);
+	if (rc != ABT_SUCCESS) {
```
Comment (Nasf-Fan) [inline src/vos/tests/vos_cmd.c:761]: ABT_finalize() before return?

File: src/vos/tests/vos_cmd.c [hunk 2]
```diff
@@ -0,0 +1,1064 @@
...
+static int
+punch_key(struct cmd_info *cinfo)
+{
+	int             rc = 0;
+	daos_key_t      dkey;
+	daos_unit_oid_t oid;
+
+	set_oid(&oid);
+
+	/* Set up dkey and akey */
```
Comment (Nasf-Fan) [inline src/vos/tests/vos_cmd.c:381]: Only dkey is set for this case.

**Review by Nasf-Fan — CHANGES_REQUESTED** [review]

File: src/vos/tests/vos_cmd.c [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+	printf("\t--punch, -p       <key>                  Punch dkey\n");
+	printf("\t--randomize, -z   <key>@<start>-<length> Randomize I/O over a range\n");
+	printf("\t--iterate, -i                            Iterate\n");
+	printf("\t--aggregate, -a                          Aggregate\n");
+	printf("\t--discard, -r                            Discard writes\n");
+	exit(0);
+}
+
+static int
+alloc_pool(const char *name, bool create, struct known_pool **pool)
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/vos/tests/vos_cmd.c:140]: I've changed it but pool_create is exclusive so it's based on that.

File: src/vos/tests/vos_cmd.c [hunk 4]
```diff
@@ -0,0 +1,1064 @@
...
+		return 1;
+
+	dts_create_config(test_name, "VOS command line %s", args.a_cfg);
+
+	if (abit_start() != 0) {
+		D_ERROR("Failed to init abt\n");
+		return 1;
+	}
+
+	for (c = 0; c < sizeof(write_buf); c++)
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply src/vos/tests/vos_cmd.c:958]: Not sure what your question is?  It's creating an array that has abcdefghijklmnopqrstuvwxyz repeated over and over.  It's helpful in the output to visually inspect the resulting merged extents. Initially, I set the whole thing to one character.

[the remark this answers is not in the document]
Comment (Nasf-Fan) [reply src/vos/tests/vos_cmd.c]: The write_buf is define as following:

#define IO_SIZE (16 * 1024 * 1024)
static char write_buf[IO_SIZE];

Then here, the for() block will repeat IO_SIZE times to initialize the write_buf, that seems quite inefficient.

**Comment by jolivier23** [comment]
ping reviewers

File: src/vos/tests/vos_cmd.c [hunk 5]
```diff
@@ -0,0 +1,1072 @@
...
+			shares[cursor++] = op;
+	/** The values defined in FOREACH_OP are percentages and should add up to 100 which should
+	 *  also be equal to the size of the array shares.should add up to the number of entries
+	 *  in the array.  If we change it to some other shares, they should still add up to the
+	 *  number in the array.
+	 */
+	D_ASSERT(cursor == ARRAY_SIZE(shares));
+
+	for (c = 0; c < sizeof(write_buf); c++)
+		write_buf[c] = (c % 26) + 97;
```
Comment (Nasf-Fan) [inline src/vos/tests/vos_cmd.c:237]: My concern is that the for section will repeat for 16 * 1024 * 1024 times to initialize the write_buf. It is some inefficient.

**Review by Nasf-Fan — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]

File: src/vos/tests/vos_cmd.c (hunk 5 above)
Comment (jolivier23) [reply src/vos/tests/vos_cmd.c:237]: Yeah, it's a one time thing and takes 0.07 seconds on my 7 year old machine so not worth optimizing.

**Pull request merged**